keps/sig-auth/20190926-dynamic-audit-policy.md

```diff
@@ -0,0 +1,588 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The proposal is to enhance the [Dynamic Audit Backend](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#dynamic-backend) feature with **audit policy** that has the following characteristics:
+- policies are defined with the `auditregistration.k8s.io/v1alpha1` [`AuditSink`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#auditsink-v1alpha1-auditregistration) API, which has a [`Policy`](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/auditregistration/types.go#L77) structure for that purpose.
+- Policies define the default level of audit detail for requests to the API server.
+- A level can be assigned also to a subset of all requests that are selected with rules.
+- Rules are decoupled from policies into dedicated `auditregistration.k8s.io/v1alpha1` `AuditClass` API objects. 
+- An `AuditClass` object can be *referenced* by multiple policies.
+- Rules in `AuditClass` resources are typed, conforming to the kinds of requests to the API server (e.g. for resource, cluster resource, url, etc.).
+- Each rule is a set of selectors for properties of the request (e.g. users, verbs, namespaces, etc)
+
+The newly proposed policy model is specifically tailored to the dynamic audit backend use case, promoting separation of concerns and policy design efficiency. It retains semantic interoperability with the existing `audit.k8s.io/v1` `Policy` model.
+
+> Note, that this KEP focuses exclusively on dynamic audit policies. For other concerns related to the dynamic audit sink registration feature, please refer to the corresponding [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/0014-dynamic-audit-configuration.md).
+
+## Motivation
+
+The use of the `kube-apiserver` audit policy command-line flag `--audit-policy-file` to set cluster audit policy is suited primarily for scenarios where exclusive auditing control for the cluster provider is either sufficient or required. The user scenarios and context elaborated in the [design proposal draft](https://docs.google.com/document/d/12uYuvykipkG96EJ4PsFtvhzgMY6mmsBgX8kNLuOq_RE/edit?usp=sharing) reveal actual use cases where this approach has shortcomings. 
+
+### Multiple audit stakeholder roles
+
+In environments, such as managed clusters (e.g. GKE), there are multiple roles that are interested in controlling certain audit aspects that do not necessarily overlap, or at least not completely. When a single policy controlled only by one of the roles is not the actual goal, it is a limitation that cannot address adequately the interests of all auditing stakeholders.
+
+### Insight into cluster activities on the spot
```

Comment:
I'm worried that we're conflating too many use cases, and that's why this KEP is so long and complicated. I agree that audit logging is useful for this, but I don't think sig-auth is the right owner for it. There are a lot of different ways one could instrument a Kubernetes deployment (e.g. [distributed tracing](https://github.com/kubernetes/enhancements/pull/650)).

Reply:
> Can you think of any ways to break this problem down into smaller pieces that can be addressed independently?

Not before the kep freeze